Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -290,60 +279,73 @@ class HorizontalFlowLayout: UICollectionViewLayout {
     }
 
-    func numberOfPages() -> Int {
-        let itemsPerPage = maxVerticalItemsCount() * maxHorizontalItemsCount()
+    func numberOfPages(with bounds: CGSize) -> Int {
+        let itemsPerPage = maxVerticalItemsCount(height: bounds.height) * maxHorizontalItemsCount(width: bounds.width)
         // Sometimes itemsPerPage is 0. In this case just return 0. We dont want to try dividing by 0.
         return itemsPerPage == 0 ? 0 : Int(ceil(Double(cellCount) / Double(itemsPerPage)))
     }
 
-    override var collectionViewContentSize: CGSize {
-        let contentSize = boundsSize
-        let horizontalItemsCount = maxHorizontalItemsCount()
-        let verticalItemsCount = maxVerticalItemsCount()
+    func calculateLayout(for size: CGSize) -> (size: CGSize, cellSize: CGSize, cellInsets: UIEdgeInsets) {
+        let width = size.width
+        let height = size.height
+        guard width != 0 else {
+            return (size: CGSize.zero, cellSize: self.itemSize, cellInsets: self.insets)
+        }
+
+        let horizontalItemsCount = maxHorizontalItemsCount(width: width)
+        let verticalItemsCount = maxVerticalItemsCount(height: height)
 
         // Take the number of cells and subtract its space in the view from the height. The left over space is the white space.
         // The left over space is then devided evenly into (n + 1) parts to figure out how much space should be inbetween a cell
-        var verticalInsets = (contentSize.height - (CGFloat(verticalItemsCount) * itemSize.height)) / CGFloat(verticalItemsCount + 1)
-        var horizontalInsets = (contentSize.width - (CGFloat(horizontalItemsCount) * itemSize.width)) / CGFloat(horizontalItemsCount + 1)
+        var verticalInsets = (height - (CGFloat(verticalItemsCount) * itemSize.height)) / CGFloat(verticalItemsCount + 1)
+        var horizontalInsets = (width - (CGFloat(horizontalItemsCount) * itemSize.width)) / CGFloat(horizontalItemsCount + 1)
 
         // We want a minimum inset to make things not look crowded. We also don't want uneven spacing.
         // If we dont have this. Set a minimum inset and recalculate the size of a cell
-        if horizontalInsets < minimumInsets || horizontalInsets != verticalInsets {
-            verticalInsets = minimumInsets
-            horizontalInsets = minimumInsets
-            itemSize.width = (contentSize.width - (CGFloat(horizontalItemsCount + 1) * horizontalInsets)) / CGFloat(horizontalItemsCount)
-            itemSize.height = itemSize.width
+        var estimatedItemSize = itemSize
+        if horizontalInsets < ASHorizontalScrollCellUX.MinimumInsets || horizontalInsets != verticalInsets {
+            verticalInsets = ASHorizontalScrollCellUX.MinimumInsets
+            horizontalInsets = ASHorizontalScrollCellUX.MinimumInsets
+            estimatedItemSize.width = (width - (CGFloat(horizontalItemsCount + 1) * horizontalInsets)) / CGFloat(horizontalItemsCount)
```

Comment:
Nit: For some of these CGFloat calculations it might be good to floor() them so we don't end up up in sub-pixel land.